### PR TITLE
Integrate a governance XCM origin 

### PR DESCRIPTION
**Pull request by gavofyork**

### CODE REVIEW

File: rococo-parachains/runtime/src/lib.rs [hunk 1]
```diff
@@ -302,6 +303,8 @@ pub type XcmOriginToTransactDispatchOrigin = (
 	// Native signed account converter; this just converts an `AccountId32` origin into a normal
 	// `Origin::Signed` origin of the same 32-byte value.
 	SignedAccountId32AsNative<RococoNetwork, Origin>,
+	// Xcm origins can be represented natively under the Xcm pallet's Xcm origin.
+	XcmPassthough<Origin>,
```
Comment (apopiak) [inline rococo-parachains/runtime/src/lib.rs:307]:
```suggestion
	XcmPassthrough<Origin>,
```

File: rococo-parachains/runtime/src/lib.rs [hunk 2]
```diff
@@ -64,6 +64,7 @@ use xcm_builder::{
 	AllowUnpaidExecutionFrom, ParentAsSuperuser,
 };
 use xcm_executor::{Config, XcmExecutor};
+use pallet_xcm::{XcmPassthough, EnsureXcm, IsMajorityOfBody};
```
Comment (apopiak) [inline rococo-parachains/runtime/src/lib.rs:67]:
```suggestion
use pallet_xcm::{XcmPassthrough, EnsureXcm, IsMajorityOfBody};
```

File: rococo-parachains/runtime/Cargo.toml [hunk 3]
```diff
@@ -27,6 +27,7 @@ sp-inherents = { git = "https://github.com/paritytech/substrate", default-featur
 frame-support = { git = "https://github.com/paritytech/substrate", default-features = false, branch = "master" }
 frame-executive = { git = "https://github.com/paritytech/substrate", default-features = false, branch = "master" }
 frame-system = { git = "https://github.com/paritytech/substrate", default-features = false, branch = "master" }
+pallet-assets = { git = "https://github.com/paritytech/substrate", default-features = false, branch = "master" }
```
Comment (shawntabrizi) [inline rococo-parachains/runtime/Cargo.toml:30]: std import

File: rococo-parachains/runtime/src/lib.rs [hunk 4]
```diff
@@ -376,6 +379,34 @@ impl cumulus_ping::Config for Runtime {
...
 }
 
+parameter_types! {
+	pub const AssetDeposit: Balance = 100;
+	pub const ApprovalDeposit: Balance = 1;
+	pub const StringLimit: u32 = 50;
+	pub const MetadataDepositBase: Balance = 10;
+	pub const MetadataDepositPerByte: Balance = 1;
+	pub const UnitBody: BodyId = BodyId::Unit;
+}
```
Comment (shawntabrizi) [inline rococo-parachains/runtime/src/lib.rs:389]: Even for Rococo, probably not the right units? Should at least use some kind of `DOLLARS` and stuff.

**Comment by apopiak** [comment]
So when we try something similar in statemint the message doesn't seem to be handled correctly (I'm still figuring out how to debug this better.), but I have a guess at where this is failing: Cumulus handles downward messages [here](https://github.com/paritytech/cumulus/blob/689cb16660c6f9ea956952a365ea621b778a52f8/primitives/utility/src/lib.rs#L71) by giving them the `Parent` origin and I'm wondering whether how that interferes with the origin Passthrough.

File: pallets/parachain-system/src/lib.rs [hunk 5]
```diff
@@ -485,25 +485,38 @@ impl<T: Config> Module<T> {
...
+			Self::deposit_event(RawEvent::DownwardMessagesProcessed(
+				processed_count,
+				weight_used,
+				result_mqc_head.clone(),
+				expected_dmq_mqc_head.clone(),
+			));
+
+			// After hashing each message in the message queue chain submitted by the collator, we should
+			// arrive to the MQC head provided by the relay chain.
+			assert_eq!(result_mqc_head, expected_dmq_mqc_head);
```
Comment (shawntabrizi) [inline pallets/parachain-system/src/lib.rs:516]: we could put this above the deposit event and then avoid the clone?

Comment (shawntabrizi) [reply pallets/parachain-system/src/lib.rs:516]: Also we want this panic?

File: pallets/xcm/src/lib.rs [hunk 6]
```diff
@@ -36,16 +40,68 @@ pub mod pallet {
...
+	}
+
+	/// For an incoming downward message, this just adapts an XCM executor and executes DMP messages
+	/// immediately up until some `MaxWeight` at which point it errors. Their origin is asserted to be
+	/// the Parent location.
+	impl<T: Config> DownwardMessageHandler for Pallet<T> {
+		fn handle_downward_message(msg: InboundDownwardMessage) -> Weight {
+			let id = sp_io::hashing::twox_64(&msg.msg[..]);
+			let msg = VersionedXcm::<T::Call>::decode(&mut &msg.msg[..])
+				.map(Xcm::<T::Call>::try_from);
```
Comment (apopiak) [inline pallets/xcm/src/lib.rs:85]: the weight for this is left out intentionally, I assume?

File: pallets/parachain-system/src/lib.rs (hunk 5 above)
Comment (gavofyork) [reply pallets/parachain-system/src/lib.rs:516]: yeah - it was relying on an indirect panic before. this just makes it explicit.

File: pallets/xcm/src/lib.rs (hunk 6 above)
Comment (gavofyork) [reply pallets/xcm/src/lib.rs:85]: correct. the weight returned is the unknowable weight. stuff for bookkeeping, dispatch, decode and conversion can be benchmarked normally.

**Comment by gavofyork** [comment]
@apopiak @shawntabrizi approvals?

File: pallets/parachain-system/src/lib.rs (hunk 5 above)
Comment (gavofyork) [reply pallets/parachain-system/src/lib.rs:516]: also done in this PR: https://github.com/paritytech/cumulus/pull/413

**Review by apopiak — APPROVED** [review]

**Pull request merged**